File: find-the-duplicate-number/find-the-duplicate-number.cpp
class Solution {
public:
int findDuplicate(vector<int>& nums) {
    int slow = nums[0];
    int fast = nums[0];

    // Find the intersection point of two pointers
    do {
        slow = nums[slow];
        fast = nums[nums[fast]];
    } while (slow != fast);

    // Move one pointer to the start and another at the intersection point
    slow = nums[0];
    while (slow != fast) {
        slow = nums[slow];
        fast = nums[fast];
    }

    return slow;
}

};

File: 287-find-the-duplicate-number/find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int n = nums.length;
       Set<Integer> s = new HashSet<>();
       for(int i =0;i<n;i++){
        if(s.contains(nums[i]))return nums[i];
        s.add(nums[i]);
       }

    return 0;
        
    }
}
